src/software/simulated_tests/tbots_test_runner.py

```diff
@@ -0,0 +1,143 @@
+        blue_full_system_proto_unix_io,
+        yellow_full_system_proto_unix_io,
+        gamecontroller,
+    ):
+        """Initialize the TestRunner. 
+        
+        :param test_name: The name of the test to run
+        :param blue_full_system_proto_unix_io: The blue full system proto unix io to use
+        :param yellow_full_system_proto_unix_io: The yellow full system proto unix io to use
+        :param gamecontroller: The gamecontroller context managed instance 
+
+        """
+
+        self.test_name = test_name
+        self.blue_full_system_proto_unix_io = blue_full_system_proto_unix_io
+        self.yellow_full_system_proto_unix_io = yellow_full_system_proto_unix_io
+        self.gamecontroller = gamecontroller
+        self.world_buffer = ThreadSafeBuffer(buffer_size=20, protobuf_type=World)
+
+        self.last_exception = None
+
+        self.ssl_wrapper_buffer = ThreadSafeBuffer(
+            buffer_size=1, protobuf_type=SSL_WrapperPacket
+        )
+        self.robot_status_buffer = ThreadSafeBuffer(
+            buffer_size=1, protobuf_type=RobotStatus
+        )
+
+        self.blue_full_system_proto_unix_io.register_observer(
+            SSL_WrapperPacket, self.ssl_wrapper_buffer
+        )
+        self.blue_full_system_proto_unix_io.register_observer(
+            RobotStatus, self.robot_status_buffer
+        )
+
+        self.blue_full_system_proto_unix_io.register_observer(World, self.world_buffer)
+
+        self.timestamp = 0
+        self.timestamp_mutex = threading.Lock()
+
+    def send_gamecontroller_command(
+        self,
+        gc_command: proto.ssl_gc_state_pb2.Command,
+        isBlue: bool,
+        final_ball_placement_point=None,
+    ):
+        """sends a gamecontroller command that is to be broadcasted to the given team
+
+        param gc_command: the gamecontroller command
+        param isBlue: whether the command should be sent to the blue team
+        final_ball_placement_point: where to place the ball in ball placement
+
+        Raises:
+            NotImplementedError: 
+        """
+        if isBlue:
+            self.gamecontroller.send_ci_input(
+                gc_command=gc_command,
+                team=Team.BLUE,
+                final_ball_placement_point=final_ball_placement_point,
+            )
+        else:
+            self.gamecontroller.send_ci_input(
+                gc_command=gc_command,
+                team=Team.YELLOW,
+                final_ball_placement_point=final_ball_placement_point,
+            )
+
+    def set_tactics(
+        self, tactics: AssignedTacticPlayControlParams, isBlue: bool,
+    ):
+        """Overrides current AI tactic for the given team
+
+        param tactic: the tactic params proto to use
+        param isBlue: whether the play should be applied to the blue team
+
+        Raises:
+            NotImplementedError
+        """
+        raise NotImplementedError("abstract method set_tactics called from base class")
```

Comment:
The implementations of this function in both `FieldTestRunner` and `SimulatorTestRunner` seems to be identical. Could probably pull the implementation up to here. Some of the other abstract methods below I think might be able to be pulled up to this class as well.